test/testdata/resolver/final_module.rb

```diff
@@ -0,0 +1,64 @@
+# typed: true
+# disable-fast-path: true
+
+class FinalClass
+  extend T::Helpers
+  final!
+end
+
+module FinalModule
+  extend T::Helpers
+  final!
+end
+
+module NonFinalModule; end
+
+class BadInherit < FinalClass # error: `FinalClass` was declared as final and cannot be inherited by `BadInherit`
+end
+
+module BadInclude
+  include NonFinalModule
+  extend NonFinalModule
+  include FinalModule # error: `FinalModule` was declared as final and cannot be included by `BadInclude`
+end
+
+module BadExtend
+  include NonFinalModule
+  extend NonFinalModule
+  extend FinalModule # error: `FinalModule` was declared as final and cannot be extended by `BadExtend`
```

Comment:
Given the location of the error message, I might change the message itself to say "extended **within**" or "extended **in**" but that's not a blocker.